src/FNA3D_Driver_Vulkan.c

```diff
@@ -6808,45 +6808,26 @@ static void VULKAN_INTERNAL_SetBufferData(
 	VulkanBuffer *vulkanBuffer = (VulkanBuffer*) buffer;
 	uint32_t prevIndex;
 	uint8_t allocateResult;
-	uint32_t i;
 
 	#define CURIDX vulkanBuffer->currentSubBufferIndex
 	#define SUBBUF vulkanBuffer->subBuffers[CURIDX]
 
 	prevIndex = CURIDX;
 
-	if (options != FNA3D_SETDATAOPTIONS_NOOVERWRITE)
-	{
-		/* If buffer has not been bound this frame, set the first unbound index */
-		if (!vulkanBuffer->bound)
-		{
-			for (i = 0; i < vulkanBuffer->subBufferCount; i += 1)
-			{
-				if (vulkanBuffer->subBuffers[i]->bound == -1)
-				{
-					break;
-				}
-			}
-			CURIDX = i;
-		}
-	}
-
-	/*
-	 * If buffer was bound and options is NONE or DISCARD
-	 * find the next available unbound sub-buffer
+	/* If NONE or DISCARD is set, we check if the buffer was bound either this frame or the previous frame.
+	 * If so, we start at sub-buffer 0 and increment the index until we find a sub-buffer that is unbound.
+	 * Otherwise we use the current sub-buffer index.
 	 */
-	if (vulkanBuffer->bound)
+	if ((options != FNA3D_SETDATAOPTIONS_NOOVERWRITE && (vulkanBuffer->bound || vulkanBuffer->boundSubmitted)))
```

Comment:
Had a bit of trouble parsing this one:

```suggestion
	if (	options != FNA3D_SETDATAOPTIONS_NOOVERWRITE &&
		(vulkanBuffer->bound || vulkanBuffer->boundSubmitted)	)
```